### PR TITLE
add functionality to search parent dirs for git repo

### LINKED ISSUE #23: Search parent directories for .git
The plugin currently expects the .git directory to be in the project root, but git itself will search up through parent directories.

### PR DESCRIPTION
**Pull request by jmcampanini**
fixes #23 
closes #18

### CODE REVIEW

File: src/main/groovy/com/palantir/gradle/gitversion/GitVersionPlugin.groovy [hunk 1]
```diff
@@ -46,5 +46,20 @@ class GitVersionPlugin implements Plugin<Project> {
...
 
+    private File findRootGitDir(File currentRoot) {
+        File gitDir = new File(currentRoot, '.git')
+
+        if (gitDir.exists()) {
+            return gitDir
+        }
+
+        // do not follow to the root directories, return non-existing File object
+        if (currentRoot.parent == '/' || currentRoot.parent == '~') {
```
Comment (markelliot) [inline src/main/groovy/com/palantir/gradle/gitversion/GitVersionPlugin.groovy]: does this work on windows?

Comment (sfackler) [inline src/main/groovy/com/palantir/gradle/gitversion/GitVersionPlugin.groovy]: File::getParentFile will return null if there is no parent. That might be a more robust check.

File: src/main/groovy/com/palantir/gradle/gitversion/GitVersionPlugin.groovy [hunk 2]
```diff
@@ -46,5 +46,20 @@ class GitVersionPlugin implements Plugin<Project> {
...
 
+    private File findRootGitDir(File currentRoot) {
+        File gitDir = new File(currentRoot, '.git')
+
+        if (gitDir.exists()) {
+            return gitDir
+        }
+
+        // stop at the root directory, return non-existing File object
+        if (currentRoot.parentFile == null) {
```
Comment (jmcampanini) [inline src/main/groovy/com/palantir/gradle/gitversion/GitVersionPlugin.groovy:58]: 1. switched to using `File` methods
2. detecting root by using `File::parentFile == null`

this should be safe on windows. don't have a win machine to test on though :-\

File: src/test/groovy/com/palantir/gradle/gitversion/GitVersionPluginTests.groovy [hunk 3]
```diff
@@ -52,6 +52,27 @@ class GitVersionPluginTests extends Specification {
...
+            version gitVersion()
+        '''.stripIndent()
+        Git git = Git.init().setDirectory(rootFolder).call();
+
+        when:
+        // will build the project at projectDir
+        BuildResult buildResult = with('printVersion').build()
+
+        then:
+        buildResult.output.contains(':printVersion\nunspecified\n')
```
Comment (markelliot) [inline src/test/groovy/com/palantir/gradle/gitversion/GitVersionPluginTests.groovy]: sorry I didn't catch this earlier, would like to actually have a commit to test we get a real version (as that guarantees there's actually a git folder there)

**Comment by jmcampanini** [comment]
let me know if this needs to be squashed before merging

**Comment by markelliot** [comment]
yep, please squash and then I'll merge + release.

**Comment by jmcampanini** [comment]
squashed :-D

**Pull request merged**